isisd/isis_spf_delay.c

```diff
@@ -0,0 +1,432 @@
+/*
+ * This is an implementation of the IETF SPF delay algorithm 
+ * as explain in draft-ietf-rtgwg-backoff-algo
+ *
+ * Module name: isis_spf_delay.c
+ * Version:     0.1
+ * Created:     25-01-2017 by S. Litkowski 
+ * Copyright (C) 2017 Orange Labs http://www.orange.com/
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
+ * MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+#include <sys/time.h>
+#include <stdlib.h>
+
+#include "thread.h"
+#include "memory.h"
+#include "prefix.h"
+
+#include "isisd/isis_constants.h"
+#include "isisd/isisd.h"
+#include "isisd/isis_adjacency.h"
+#include "isisd/isis_spf.h"
+#include "isisd/isis_spf_delay.h"
+
+const char       *isis_spf_delay_states[3] = { "QUIET", 
+					       "SHORT_WAIT", 
+					       "LONG_WAIT"};
+
+DEFINE_MTYPE_STATIC(ISISD, ISIS_SPF_DELAY_REINIT, "ISIS spf delay reinit")
+DEFINE_MTYPE_STATIC(ISISD, ISIS_SPF_DELAY_AF, "ISIS spf delay af")
+DEFINE_MTYPE_STATIC(ISISD, ISIS_SPF_DELAY, "ISIS spf delay")
+
+
+int
+isis_spf_delay_event_ttl_expires(struct thread *thread) {
+
+  int level = 0;
+  struct isis_spf_delay_reinit    *reinit;
+  struct isis_area        *area;
+  int family = 0;
+
+  reinit = THREAD_ARG(thread);
+  
+  area = reinit->area;
+  level = reinit->level;
+  family = reinit->family;
+
+  if (family == AF_INET) {
```

Comment:
That's true, the issue that I had when I wrote the code, is that Quagga implemented two scheduling functions isis_spf_schedule6 and isis_spf_schedule which is weird because as you mentionned IS-IS SPF runs per topology not just per AF. So I tried to mimic this code philoshophy when building the structures.

If we keep the two scheduling functions, it will be hard to maintain timers correctly as the two scheduling functions will try to play with the timers while only one should do.

So I do not really see how to address it easily (without touching the initial scheduling code). Any opinion ?
If IPv4 and IPv6 are activated without MT (as today), IPv4 and IPv6 timer structure will evolve in exactly the same way as isis_spf_schedule and isis_spf_schedule6 are called at the same time.